src/mbgl/mtl/buffer_resource.cpp

```diff
@@ -90,13 +95,42 @@ void BufferResource::update(const void* data, std::size_t updateSize, std::size_
     updateSize = std::min(updateSize, size - offset);
 
     if (updateSize > 0) {
+        auto& stats = context.renderingStats();
         if (buffer) {
-            if (auto* content = static_cast<uint8_t*>(buffer->contents())) {
-                std::memcpy(content + offset, data, updateSize);
+            if (auto* const content = static_cast<uint8_t*>(buffer->contents())) {
```

Comment:
I'd be curious if calling `buffer->contents()` is doing us any favors here. Mapping a buffer to host memory is usually a heavier operation.

The [docs](https://developer.apple.com/documentation/metal/mtlbuffer/1515716-contents) don't make it immediately obvious if this is a full map operation, but I would assume it is.

Reply:
Because we're forced to use `MTLResourceStorageModeShared`, I think this is just CPU memory that's mapped when the command buffer is committed, but it could be doing some mapping here.  We'll just have to measure it.

Reply:
I don't think there's any way for this API to understand the context of a call to `contents()`. It could be for a read only purpose or the caller could try and write to the memory it points at. In the latter case a transfer has to happen between host and GPU. And then thinking to the case of a compute shader that writes to a buffer, the API would also possibly need to transfer the other way when called as well. This is essentially a full buffer map then, if we cover those situations. Definitely measure it, I'm interested to see how responsive it is on the happy path where it doesn't need to do anything.

Reply:
we could skip calling `buffer->contents()` if we know we're going to replace it entirely, if calling it is expensive.

Reply:
I re-did the buffer update logic, please have another look.